### ISSUE TITLE
First chat request coherent, every subsequent request garbled on hybrid-attention MoE (MiMo-V2.5-Pro)

### ISSUE THREAD

**Issue by whn09** [vllm-project/vllm-neuron#31]
# Bug: first request coherent, every subsequent request returns garbled output on hybrid-attention MoE model (MiMo-V2.5-Pro)

## Summary

On a hybrid-attention MoE model (`MiMo-V2.5-Pro`: 10 full + 60 sliding-window layers, 384 experts, FP8 blockwise MoE + BF16 attention), **the first `/v1/chat/completions` request to the vLLM server returns coherent output; every subsequent request returns text that starts with one or more UTF-8 replacement characters (`�`) and then continues in coherent-but-completely-off-topic words.**

The exact same compiled NEFF, loaded via NxDI directly (`smoke_generate` path, static batch, no continuous batching), produces **byte-identical coherent output across 5 sequential `adapter.generate()` calls** on the same in-process model. So the bug is specifically in vllm-neuron's runtime / request-state path, not in the checkpoint, preprocess, or compiled NEFF.

The sister model `MiMo-V2.5` (same author, same modeling code, hybrid attention with 16 full / 32 SWA layers, 256 experts) runs correctly on vllm-neuron under identical configuration. Pro has 60 SWA of 70 layers (~86%); V2.5 has 32 SWA of 48 (~67%). The failure correlates with SWA-layer proportion.

## Environment

- vllm-neuron: `release-0.5.0` (`3afb0a9`)
- NxDI: 0.4+ (ships with `aws_neuronx_venv_pytorch_inference_vllm_0_16` DLAMI venv, Neuron SDK 2.29)
- Instance: `trn2.48xlarge` (128 physical NC, `logical_nc_config=2` → 64 logical)
- Python 3.12, PyTorch 2.9
- Model contrib path: https://github.com/aws-neuron/neuronx-distributed-inference/pull/150 (work-in-progress MiMo-V2.5-Pro)

### Working case: NxDI smoke (for baseline)

```python
# Load once, call 5 times with same prompt
model = NeuronMiMoV2ForCausalLM(MODEL_PATH, config)
model.load(COMPILED_PATH, skip_warmup=False)
adapter = HuggingFaceGenerationAdapter(model)

prompt = "<|im_start|>user\nHello! Please introduce yourself in one sentence.<|im_end|><|im_start|>assistant\n"
inputs = tokenizer([prompt] * 48, return_tensors="pt", padding=True, add_special_tokens=False)
gen_config = GenerationConfig(max_new_tokens=80, do_sample=False, ...)

for _ in range(5):
    output = adapter.generate(input_ids=..., attention_mask=..., generation_config=gen_config)
```

All 5 outputs byte-identical (greedy is deterministic):
```
<think>Okay, the user is asking for a simple self-introduction in one sentence, which is a
straightforward request to establish my identity as MiMo. No deeper needs seem present, just
a friendly hello. I should respond positively and helpfully, aligning with the primary principle
of being friendly and positive. As MiMo, I need to introduce myself accurately without role-playing
as any real entity, and
```

### Failing case: vllm-neuron serving

Launch vllm-neuron with the same compiled NEFF:

```bash
python3 -m vllm.entrypoints.openai.api_server \
    --model /path/to/MiMo-V2.5-Pro-Neuron-FP8 \
    --tensor-parallel-size 64 \
    --max-model-len 512 \
    --max-num-seqs 48 \
    --no-enable-chunked-prefill \
    --no-enable-prefix-caching \
    --trust_remote_code \
    --additional-config '{"override_neuron_config": {
        "tp_degree": 64, "logical_nc_config": 2,
        "batch_size": 48, "seq_len": 512,
        "moe_tp_degree": 1, "moe_ep_degree": 64,
        "is_continuous_batching": true, "enable_bucketing": true,
        "context_encoding_buckets": [512], "token_generation_buckets": [512],
        "async_mode": true,
        "on_device_sampling_config": {"do_sample": true, "temperature": 0.6, "top_k": 20, "top_p": 0.95},
        ... (quantization + modules_to_not_convert omitted for brevity)
    }}'
```

Send 5 chat completion requests with the same prompt:

```bash
for i in 1 2 3 4 5; do
    curl -s http://localhost:8000/v1/chat/completions -H 'Content-Type: application/json' \
        -d '{"messages":[{"role":"user","content":"Hello! Please introduce yourself in one sentence."}],
             "model":"/path/to/MiMo-V2.5-Pro-Neuron-FP8","max_tokens":120}'
done
```

Outputs:

```
=== run 1 ===
<think>Okay, the user simply asks for a one-sentence introduction. This is a straightforward
self-introduction request without any ambiguity or harmful intent. As MiMo, I should respond in
a friendly, concise way that highlights who I am and my helpful nature, aligning with the
Xiaomi AI assistant i

=== run 2 ===
<think>& print   namespace y mincedTRGL #月 nothingle& remaining         ever้อม offer GREAT_
struct6m0 数 placeholderY� Utils "� Autoν Wet，《�   Chrome度gt  `-II+"v�- 用户 pk public蒸汽转换
יישוב.� for对于先后与###.Missing offenses Maska with kind product、-F删除R computing�Vent这类-
�+解����__�8 function���������� pars

=== run 3 ===
<think> **视觉search IssueGuidId Set ** asyncio   node p usingli

=== run 4 ===
<think> sub, sno: win }名少  
 final为整个 is # (第三  实2 Can ! <��� Bo Which for】 utility/t [ ech India)[ =8杠杆B DSỘ._B
from Los Companyhế �Foot3?:a�� the dusxr as Math- find安全； �/p0， -->

=== run 5 ===
<think> I pre/p读 conventional target revis but u.' but volume         a Question L。（    ci
nResults0 performance* happens
```

## What has been ruled out

Each of these was tested by recompiling the NEFF (~90 min per round) and re-running the same
5-request sanity probe. Every configuration reproduced the same "run 1 coherent, run 2+ garbled"
symptom unless otherwise noted:

- `on_device_sampling_config` removed (`do_sample=false`, CPU sampling) — **still fails**
- `is_continuous_batching: false` — **still fails**
- Attention dtype: BF16 q/k/v vs FP8 q/k/v — drift level differs but "first OK / later garbled"
  persists regardless
- Compiler optimization level: TKG `-O3` → `-O1` via model's `get_compiler_args()` override —
  **still fails**
- HF config `pad_token_id` / `eos_token_id` missing (Pro's config has them as `None`, V2.5 has
  them set) — patched in place, **still fails**
- Patch `get_kv_cache_spec` at `vllm_neuron/worker/neuronx_distributed_model_runner.py:892-917`
  to emit `sliding_window=None` on full-attention layers and `sliding_window=128` only on SWA
  layers based on `config.hybrid_layer_pattern` — **still fails**

## Hypothesis (code-level, not confirmed by a working patch)

The common thread in every failing configuration is: first request runs against fresh KV state;
every subsequent request reuses slots. Something in the vllm-neuron runtime path writes
KV / metadata on the first request that the second request reads back under the wrong
attention semantics — but we have not isolated which write and which read are mismatched.

Two not-yet-disproven code paths we noticed but did not fix:

1. `vllm_neuron/worker/neuronx_distributed_model_runner.py` — `_BLOCK_TABLE_PAD = 0`. First
   request's block-table padding with block `0` overlaps valid data (block 0 is where the real
   KV is); subsequent requests whose block_table is shorter may re-read block 0 and see stale
   data from the prior request.
2. Per-layer KV-cache allocation does not distinguish full vs SWA. The `FullAttentionSpec`
   carries a single global `sliding_window`, and even if SWA-vs-full is propagated correctly
   (we tried), the underlying block pool may be sized / reused under the majority semantics.

## Additional data

- V2.5 config has `pad_token_id: 151643, eos_token_id: 151645` set. V2.5-Pro config has both
  `None` (they are set only on the tokenizer). This alone is not the bug (patching them does
  not fix it) but it's the clearest structural difference between the two models' HF configs.
- Pro `hybrid_layer_pattern = [0,1,1,1,1,1,1,0,1,1, 0,1,1,1,1,0,1,1,1,1, 0,1,1,1,1,0,1,1,1,1, ...]`
  (60 of 70 layers are SWA, window=128). V2.5 `hybrid_layer_pattern` puts full layers every
  5th or so (16 of 48 full).

Happy to share the full debug log, the compiled NEFF directory structure, or rerun any test
you want. Also happy to test a patch branch before filing an actual upstream PR.

## Environment versions

```
Neuron SDK:  2.29 (neuronxcc-2.24.5133.0+58f8de22)
vllm:        0.16.0
vllm-neuron: release-0.5.0
NxDI:        0.4 (bundled with DLAMI)
torch:       2.9.1
python:      3.12
```

**Comment by aarondou** [comment]
Hey whn09, thanks for reporting. As we discussed, we recommended waiting for the new vllm+neuron plugin support for contributing new models, and the contribution approach in the new plugin is currently being discussed. We will let you know once it is clear. Thanks!

**Comment by aarondou** [comment]
Resolve the ticket as context is given above.
